### PR TITLE
various optimization on memory allocation

### PR DESCRIPTION
**Pull request by fs-eire**
this change reduces unnecessary memory copies in various places.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/controlflow/if.cc [hunk 1]
```diff
@@ -139,6 +142,24 @@ Status IfImpl::Initialize() {
...
+    // alternatively we could track implicit inputs on a per-attribute basis in the node, but that
+    // would make that tracking a bit more complicated.
+    int idx;
+    if (ort_value_name_idx_map.GetIdx(entry.first, idx).IsOK()) {
+      subgraph_feed_names_.push_back(entry.first);
+      subgraph_feeds_mlvalue_idxs_.push_back(idx);
+    }
+  }
+
   auto status = AllocateOutputTensors();
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/controlflow/if.cc]: Previously this was done in CreateFeedsFetchesManager which was called once and then cached.

Now you're doing this every time the If operator's Compute method is called. 

Isn't that worse?

File: onnxruntime/core/framework/feeds_fetches_manager.h [hunk 2]
```diff
@@ -38,11 +38,17 @@ struct FeedsFetchesInfo {
   // set the ort_value_idxs for the current values in feed_names and output_names
   Status SetMLValueIdxs(const OrtValueNameIdxMap& ort_value_name_idx_map);
 
-  std::vector<std::string> feed_names;
-  std::vector<std::string> output_names;
+  FeedsFetchesInfo(FeedsFetchesInfo&& other) = default;
+  FeedsFetchesInfo& operator=(FeedsFetchesInfo&&) = default;
+
+  const std::vector<std::string>& feed_names;
+  const std::vector<std::string>& output_names;
```
Comment (skottmckay) [inline onnxruntime/core/framework/feeds_fetches_manager.h]: Do we need to have references here tying this class to the creator's lifetime, or would it have been sufficient just to have the fix to use std::move when initializing feeds_fetches_info_ in the FeedsFetchesManager ctor?

File: onnxruntime/core/framework/execution_frame.h [hunk 3]
```diff
@@ -91,7 +91,7 @@ class IExecutionFrame {
   // Input and Output values are passed in by executors
   std::vector<OrtValue> all_values_;
 
-  const std::vector<int> fetch_mlvalue_idxs_;
+  const std::vector<int>& fetch_mlvalue_idxs_;
 };
```
Comment (skottmckay) [inline onnxruntime/core/framework/execution_frame.h]: Was the copy of vector<int> a significant enough perf issue is warrant using a reference here? In the profiling I've done I've never noticed it (vs things like the cost of setting up the memory pattern planner in the ExecutionFrame ctor which could be reworked)

**Comment by pranavsharma** [comment]
As per the PR guidelines, please indicate how much improvement you saw after these fixes.

**Comment by fs-eire** [comment]
Close this PR.

Some lines are moved to #1392. The other part need to be redo in an appropriate design.

**Pull request closed without merging**